### DISCUSSION TITLE
Single chassis multiple devices

### DISCUSSION THREAD

**Discussion by royaljeffr** [netbox-community/netbox#9578] [General]
I have purchased a [TrueNAS M-40](https://www.openstoragenas.com/TrueNAS-M40.asp) that is a 4U server with 2 dual controllers.  Essentially 2 1U servers within the same chassis but has a shared IP address.  Is the preferred Netbox way to make a Device Type and a bunch of Components or create a Virtual Chassis and add Members/Devices?

**Comment by candlerb** [comment]
If these are dual controllers in an active/standby mode, where you can't login individually to the two controllers, and they are just cards in a slot, then I'd model the whole thing as a single Device, with Inventory Items for the two controller cards.

If the controller cards have their own components (e.g. network ports or console ports) then you could make them Modules in a Module bay (this requires Netbox 3.2); but even without this, you can always add components directly to the Device itself.

If the controller cards are completely independent entities, like blade servers in a blade chassis, or two completely independent firewalls, then you could model them as two separate devices inside a parent device.  You'd then probably use FHRP for the shared IP address.  But I don't think this applies here; in any case you'd have the problem of which device the hard drives belong to as Inventory Items.

I don't think virtual chassis is a good match for your use case.  This is intended for when there are two physically separate devices, with their own rack positions, but are managed as a single logical device - e.g. a switch stack with ports `Gi1/0/[1-24]` on one device and `Gi2/0/[1-24]` on another device.

**Reply by royaljeffr** [reply]
Thank you for the advice.  After reading your suggestion I did a combo of what you are talking about. 

I created the chassis as a device with a couple interfaces and 26 device bays.  Then I created each of the controllers as devices and assigned them to the chassis as a parent/child.  This seems allows me to use the shared (virtual) IPs on the chassis and add all the disks directly to the chassis in the device bays while allowing me to document the specs on each of the controllers independently.

**Reply by candlerb** [reply]
Disks are not Devices.  Devices are fully independent systems with their own management IP addresses.  Disks should be added as Inventory Items under the Device.

I don't think that making the controller card devices is right either - they should be either inventory items or modules.

Of course it's free software and you can use it in whatever way you like.  But if this were a case of line cards in a chassis router or switch, you would get definite problems modelling them as "devices" (e.g. inability to see links directly on the parent device, and inability to make LAGs between line cards)

**Reply by royaljeffr** [reply]
My thought was if I created Device Bays for each Disk I could then know which slot each HDD is really in.  If I use Inventory I don't have that option unless I add a custom field.

In regards to the controller cards they are bit more than just a card - they have cpu, memory , nics, etc..  Like in your example, more like a blade server.

I'm trying to the learn the Netbox way :).

**Reply by candlerb** [reply]
You can name the items in the inventory by their slot position, e.g. disk0 ... disk23

Controller cards which are just failover aren't really independent devices.  When you login to the device, you're not logging into a specific controller card, you're logging into the device.  You said there was a single IP address: therefore this is the address of the device, not one of the controller cards.  (Of course, this address is *served* by whichever controller card is active).

**Comment by mtinberg** [comment]
Module Bays and Modules in Netbox 3.2 are probably what you want for the disks then, they work like Device Bays/Devices but for things with augment the overall Device rather than being an independent Device themselves, uplink modules and power supplies that add ports to the overall Device are other things Modules are used for, otherwise Inventory Items with a Custom Field for position would work too as you said.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by royaljeffr** [reply]
I really like that Modules in 3.2 for my HDD's.  Thank you.
